### PR TITLE
Adds MAUI extension steps : head projects, templates, project property transforms, source code updates

### PR DESCRIPTION
**Pull request by Sweekriti91**
Looking for feedback on the code I added to handle the Multiple TFms that get added to a .NET MAUI head project. I added a component check + log warning to the IUpgradeReadyCheck to allow MAUI projects *only* 

I also added a new MSBuild property called `ProjectPropertyElements`, implementation can be found [here](https://github.com/dotnet/upgrade-assistant/blob/swsat/maui-tmpts-source/src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/MSBuildProject.PropertyElement.cs). This is for the unevaluated project properties (of which there are many in legacy Xamarin projects). These properties have new mappings or have new values which is why I needed to add this. I followed the interface pattern. I wasn't sure how to explicitly add a test for this, if someone could let me know, I can add that. 

Fixed a small typo in this message: https://github.com/dotnet/upgrade-assistant/blob/swsat/maui-tmpts-source/src/extensions/default/Microsoft.DotNet.UpgradeAssistant.Steps.Solution/CurrentProjectSelectionStep.cs#L234

For DevOps, I added this [file](https://github.com/dotnet/upgrade-assistant/blob/swsat/maui-tmpts-source/eng/restore-toolset.ps1), @twsouthwick  made a good point to add a ReadME to explain what is going on and maybe some troubleshoot steps for the future. What would be the best place to add this information? A readme in that eng folder? 

All other changes are isolated to the maui-extension project and updates to the E2E test data files to reflect the new steps added!

Fixes:
https://github.com/dotnet/upgrade-assistant/issues/551
https://github.com/dotnet/upgrade-assistant/issues/552
https://github.com/dotnet/upgrade-assistant/issues/587
 https://github.com/dotnet/upgrade-assistant/issues/590
https://github.com/dotnet/upgrade-assistant/issues/589
https://github.com/dotnet/upgrade-assistant/issues/755
https://github.com/dotnet/upgrade-assistant/issues/588

### CODE REVIEW

File: src/extensions/maui/Microsoft.DotNet.UpgradeAssistant.Extensions.Maui/UsingXamarinEssentialsAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            var usingDirective = (UsingDirectiveSyntax)context.Node;
+
+            var namespaceName = usingDirective.Name?.ToString();
+
+            if (namespaceName is null)
+            {
+                return;
+            }
+
+            if (DisallowedNamespaces.Any(name => namespaceName.Equals(name, StringComparison.Ordinal) || namespaceName.StartsWith($"{name}.", StringComparison.Ordinal)))
```
Comment (twsouthwick) [inline src/extensions/maui/Microsoft.DotNet.UpgradeAssistant.Extensions.Maui/UsingXamarinEssentialsAnalyzer.cs:50]: There are a number of potentially unnecessary allocations here. Probably not a big deal, but at this point DisallowedNamespaces is a single entry - do you expect that to expand?

File: src/extensions/maui/Microsoft.DotNet.UpgradeAssistant.Extensions.Maui/UsingXamarinFormsAnalyzerAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.DotNet.UpgradeAssistant.Extensions.Maui
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class UsingXamarinFormsAnalyzerAnalyzer : DiagnosticAnalyzer
+    {
+        public const string DiagnosticId = "XamarinForms";
```
Comment (twsouthwick) [inline src/extensions/maui/Microsoft.DotNet.UpgradeAssistant.Extensions.Maui/UsingXamarinFormsAnalyzerAnalyzer.cs:17]: We've been using `UAXXXX` where `XXXX` is a zero-padded count

File: src/extensions/maui/Microsoft.DotNet.UpgradeAssistant.Extensions.Maui/UsingXamarinFormsAnalyzerAnalyzer.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.DotNet.UpgradeAssistant.Extensions.Maui
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class UsingXamarinFormsAnalyzerAnalyzer : DiagnosticAnalyzer
```
Comment (twsouthwick) [inline src/extensions/maui/Microsoft.DotNet.UpgradeAssistant.Extensions.Maui/UsingXamarinFormsAnalyzerAnalyzer.cs:15]: This appears to be the same as the other analyzer. Can the be combined?

**Review by twsouthwick — APPROVED** [review]
This looks pretty good. My only main concern is about the UX of allowing multiple TFMs through.... it's probably fine, but I just want to make sure we're sure of that.

I'll be OOF the next few days, so I'm approving this as there's nothing I see that should block

**Comment by Sweekriti91** [comment]
@mjrousos @sunandabalu @KSchlobohm  could I also ask for your feedback on the points Taylor raised :

1. this comment to get around MultiTargets https://github.com/dotnet/upgrade-assistant/pull/790#discussion_r687268903 

Context, MAUI Head projects (post migration, head Xamarin forms project does not have multi-target) as part of migration process, get a new TargetFrameworks added into them. I handle setting the frameworks in the extension TFM filter, and only change I made was to this IUpgradeReadyCheck to stop the error being logged in the UI

2. For the Roslyn Analyzer, I need to update the naming to UAXXX, is there a full list I can verify what the last number was so I can update my analyzers with the correct values?

File: src/components/Microsoft.DotNet.UpgradeAssistant/Checks/MultiTargetFrameworkCheck.cs [hunk 4]
```diff
@@ -29,24 +29,33 @@ public MultiTargetFrameworkCheck(ILogger<MultiTargetFrameworkCheck> logger)
...
-                return Task.FromResult(UpgradeReadiness.Ready);
+                return UpgradeReadiness.Ready;
             }
             else
             {
-                _logger.LogError("Project {Project} cannot be upgraded. Input projects must have exactly one target framework. Please see {FeatureLink} to request this feature.", project.FileInfo, FEATURE_LINK);
-                return Task.FromResult(UpgradeReadiness.NotReady);
+                if (components.HasFlag(ProjectComponents.Maui))
+                {
+                    _logger.LogTrace("Confirmed project {Project} is of type .NET MAUI Head with TFMs : ({TFM}))", project.FileInfo, tfms.First());
```
[the remark this answers is not in the document]
Comment (sunandabalu) [reply src/components/Microsoft.DotNet.UpgradeAssistant/Checks/MultiTargetFrameworkCheck.cs:51]: Is the purpose here just to log that there are Maui head projects that are multi-targeting? Coz I see that the return type is still NotReady. Although the messaging is very confusing as it does not clearly tell the customer that they can't move forward coz they are multi-targeting. I would do this instead to make it more obvious: 
` 

else{

               if (components.HasFlag(ProjectComponents.Maui))
                {
                    _logger.LogTrace("Project {Project} is of type .NET MAUI Head with multiple TFMs : ({TFM}))", project.FileInfo, tfms.First());
               }
             _logger.LogError("Project {Project} cannot be upgraded. Input projects must have exactly one target framework. Please see {FeatureLink} to request this feature.", project.FileInfo, FEATURE_LINK);
                return UpgradeReadiness.NotReady;
}

`

[the remark this answers is not in the document]
Comment (Sweekriti91) [reply src/components/Microsoft.DotNet.UpgradeAssistant/Checks/MultiTargetFrameworkCheck.cs:51]: This part happens when legacy xamarin head project is converted to Maui head project, all steps are complete and then in the finalize step, you see this Log Error that multi target is not supported, I basically wanted to put wording specifically for MAUI projects that "finalized and completed for MAUI head, TFMs are added" and not show the second message because then MAUI migration folks would be confused. Also, set it to NotReady but not Error because migration is done, no more steps are needed if project head is already MAUI style with TFMs. 

So Basically, 
if MAUI == show Information message that migration complete, and move onto next project
else == show error message that Multiple TFMs not supported.

File: src/extensions/maui/Microsoft.DotNet.UpgradeAssistant.Extensions.Maui/MauiAddProjectPropertiesStep.cs [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+                }
+            }
+
+            if (components.HasFlag(ProjectComponents.MauiiOS))
+            {
+                // following conversion mapping here : https://github.com/xamarin/xamarin-macios/wiki/Project-file-properties-dotnet-migration
+                var runtimeprops = projectproperties.GetProjectPropertyValue("MtouchArch").Distinct();
+                var runtimeIdentifierString = string.Empty;
+                int count = 0;
+                foreach (var prop in runtimeprops)
```
[the remark this answers is not in the document]
Comment (sunandabalu) [reply src/extensions/maui/Microsoft.DotNet.UpgradeAssistant.Extensions.Maui/MauiAddProjectPropertiesStep.cs:88]: +1 , there is a lot of if..else processing going on here, it makes sense to move it out to a separate method

File: src/extensions/maui/Microsoft.DotNet.UpgradeAssistant.Extensions.Maui/MauiComponentIdentifier.cs [hunk 6]
```diff
@@ -39,16 +44,33 @@ public ValueTask<ProjectComponents> GetComponentsAsync(IProject project, Cancell
                 components |= ProjectComponents.XamariniOS;
             }
 
-            var targetFrameworkNames = project.TargetFrameworks.Select(r => r.Name);
+            if (packageReferences.Any(x => x.Name.Equals(_xamarinFormsPackage, StringComparison.OrdinalIgnoreCase)) && project.TargetFrameworks.FirstOrDefault().Equals(TargetFrameworkMoniker.NetStandard20))
+            {
```
Comment (sunandabalu) [inline src/extensions/maui/Microsoft.DotNet.UpgradeAssistant.Extensions.Maui/MauiComponentIdentifier.cs:48]: do you only care about the First element being .NETStandard2.0? should this not be TargetFrameworks.Any(condition)?

File: src/extensions/maui/Microsoft.DotNet.UpgradeAssistant.Extensions.Maui/MauiPlatformTargetFrameworkUpgradeStep.cs [hunk 7]
```diff
@@ -43,50 +43,84 @@ protected override async Task<UpgradeStepApplyResult> ApplyImplAsync(IUpgradeCon
...
                 return new UpgradeStepApplyResult(UpgradeStepStatus.Failed, $"componentFlag Context property was null.");
             }
 
             var targetTfm = GetExpectedTargetFramework(componentFlagProperty);
-
-            var project = context.CurrentProject.Required();
-            var file = project.GetFile();
-
             file.SetTFM(targetTfm);
+            Logger.LogInformation("Added TFM to {TargetTFM}", targetTfm);
```
Comment (sunandabalu) [inline src/extensions/maui/Microsoft.DotNet.UpgradeAssistant.Extensions.Maui/MauiPlatformTargetFrameworkUpgradeStep.cs:76]:
```suggestion
            Logger.LogInformation("TFM set to {TargetTFM}", targetTfm);
```

File: src/components/Microsoft.DotNet.UpgradeAssistant/Checks/MultiTargetFrameworkCheck.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sunandabalu) [reply src/components/Microsoft.DotNet.UpgradeAssistant/Checks/MultiTargetFrameworkCheck.cs:51]: IsReadyCheck is call that is made at the beginning of the upgrade in the CurrentProjectSelectionStep, I don't see the readyCheck being called in the FinalizeSolutionStep. Maybe I'm missing something?

File: src/extensions/maui/Microsoft.DotNet.UpgradeAssistant.Extensions.Maui/UsingXamarinFormsAnalyzerAnalyzer.cs (hunk 3 above)
Comment (Sweekriti91) [reply src/extensions/maui/Microsoft.DotNet.UpgradeAssistant.Extensions.Maui/UsingXamarinFormsAnalyzerAnalyzer.cs:15]: yup combined it!

**Comment by Sweekriti91** [comment]
Issue to discuss the Multitarget Project Flow : https://github.com/dotnet/upgrade-assistant/issues/797

File: src/components/Microsoft.DotNet.UpgradeAssistant/Checks/MultiTargetFrameworkCheck.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Sweekriti91) [reply src/components/Microsoft.DotNet.UpgradeAssistant/Checks/MultiTargetFrameworkCheck.cs:51]: marking this resolved, after talking to you both, created this issue to further discuss the best path forward : https://github.com/dotnet/upgrade-assistant/issues/797
